jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java

```diff
@@ -217,225 +169,158 @@ public void execute(Runnable task) throws RejectedExecutionException
     @Override
     public boolean tryExecute(Runnable task)
     {
-        if (LOG.isDebugEnabled())
-            LOG.debug("{} tryExecute {}", this, task);
         if (task == null)
             return false;
 
-        // Offer will only succeed if there is a reserved thread waiting
-        boolean offered = _queue.offer(task);
-
-        // If the offer succeeded we need to reduce the size, unless it is set to -1 in the meantime
-        int size = _count.getLo();
-        while (offered && size > 0 && !_count.compareAndSetLo(size, --size))
-            size = _count.getLo();
+        ReservedThread reserved = _threads.take();
+        if (reserved != null)
+            return reserved.wakeup(task);
 
-        // If size is 0 and we are not stopping, start a new reserved thread
-        if (size == 0 && task != STOP)
-            startReservedThread();
+        startReservedThread();
 
-        return offered;
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} tryExecute failed for {}", this, task);
+        return false;
     }
 
     private void startReservedThread()
     {
-        while (true)
+        try
+        {
+            ReservedThread thread = new ReservedThread();
+            _executor.execute(thread);
```

Comment:
This feels too simplistic?

Under load, we would now start a lot of `ReservedThread`s, just to return from their `Runnable` immediately because they don't fit the `ThreadIdPool`.

Before we were accurately counting before starting a new `ReservedThread`, I'm not sure we can completely remove that logic?

Reply:
I did have a version that would only allow a single pending reserved thread and that didn't make a difference either way.   Probably because we now have a >99% hit rate mostly, so we rarely need to start a new reserved thread anyway.    

So unless we can see a real problem with starting too many reserved threads, I'd prefer less code.  i.e. lets not try to fix a problem we don't have

Reply:
>  Probably because we now have a >99% hit rate mostly

In a synthetic benchmark on a limited machine.

If `tryExecute()` is called concurrently, we obviously create a lot of platform threads to just exit them immediately.